scalafix-core/shared/src/main/scala/scalafix/internal/rule/ImplicitClassPrivateParams.scala

```diff
@@ -0,0 +1,52 @@
+package scalafix.internal.rule
+
+import scala.meta.Term.Param
+import scala.meta.{Defn, Mod}
+import scalafix.rule.RuleCtx
+import scalafix.{Patch, Rule}
+
+case object ImplicitClassPrivateParams
+    extends Rule("ImplicitClassPrivateParams") {
+
+  override def description: String =
+    "Add private access modifier to val or var parameters of implicit classes in order to prevent public access"
+
+  override def fix(ctx: RuleCtx): Patch = {
+    ctx.tree.collect {
+      case Defn.Class(mods, _, _, primCtor, _) if hasImplicitMod(mods) =>
```

Comment:
I think we can refine the condition to only apply for AnyVals that are missing `private`

```scala
        t.collect {
          case Defn.Class(
              cmods,
              _,
              _,
              Ctor.Primary(_, _, (Term.Param(pmods, _, _, _) :: Nil) :: Nil),
              Template(_, init"AnyVal" :: Nil, _, _)
              ) =>
            for {
              _ <- cmods.find(_.is[Mod.Implicit])
              if !pmods.exists(_.is[Mod.Private])
              valmod <- pmods.find(_.is[Mod.ValParam])
            } yield ctx.addLeft(valmod, "private ")
        }
```

For non-anyvals it's possible to remove the `val` parameter, `implicit class Foo(x: Int)  { .. }`. Classes that `extends AnyVal` are however forced to add the `val` modifier which makes  it easy to accidentally forget `private`

Reply:
Got it. Can you explain to me how `is()` is working here ? I noticed it but thought it only worked with tokens. Where are the instances for `Classifiable`/`Classifier` ?

Reply:
The instances are generated by the scalameta `@ast` macro annotations. It's a flexible mechanism to guard against categories of trees/tokens both via `.is[T]` or in pattern matching with `case T()`. You can build custom groups of classes like we do in scalafix https://github.com/scalacenter/scalafix/blob/00199145261b958fb12cf256773c2ed08f336359/scalafix-core/shared/src/main/scala/scalafix/util/TokenClasses.scala

Reply:
Ok thanks. One last question : your implementation ignores cases where `final` or `protected` is present. Don't we care ?